### PR TITLE
js: Add async sub for PullSubscribe to make fetch requests

### PR DESCRIPTION
**Pull request by wallyqs**
This makes the pull based subscription have their own requests handler to make fetch requests using a single wildcard subscription similar to how `nc.Request` works. Unlike `nc.Request` though, it does not use a token to identify response to requestor since the inbox gets rewritten when the JS message is delivered, and just sends the latest response to the first available requestor.

Signed-off-by: Waldemar Quevedo <wally@synadia.com>

### CODE REVIEW

File: js.go [hunk 1]
```diff
@@ -400,6 +401,34 @@ func (paf *pubAckFuture) Msg() *Msg {
...
+func (js *js) pullSubscribe(subj string) (*Subscription, error) {
+	jsi := &jsSub{js: js, pull: true}
+
+	// Similar to async request handler we create a wildcard subscription for making requests,
+	// though we do not use the token based approach since we cannot match the response to
+	// the requestor due to JS subject being remapped on delivery.  Instead, we just use an array
+	// of channels similar to how ping/pong interval is handled and send the message to the first
+	// available requestor via a channel.
+	sha := sha256.New()
+	sha.Write([]byte(nuid.Next()))
```
Comment (derekcollison) [inline js.go:415]: If using sha can just pull crypto.Rand() as well. Not sure if this was copied.

File: js.go [hunk 2]
```diff
@@ -400,6 +401,34 @@ func (paf *pubAckFuture) Msg() *Msg {
 	return paf.msg
 }
 
+// pullSubscribe creates the wildcard subscription used per pull subscriber
+// to make fetch requests.
+func (js *js) pullSubscribe(subj string) (*Subscription, error) {
+	jsi := &jsSub{js: js, pull: true}
+
+	// Similar to async request handler we create a wildcard subscription for making requests,
```
Comment (derekcollison) [inline js.go:409]: Is this for all pull subscribers? There should be one of these for each pull subscriber instance I believe.

File: js.go [hunk 3]
```diff
@@ -810,7 +839,15 @@ type nextRequest struct {
 
 // jsSub includes JetStream subscription info.
 type jsSub struct {
-	js       *js
+	js *js
+
+	// To setup request mux handler for pull subscribers.
+	mu    sync.RWMutex
```
Comment (derekcollison) [inline js.go:845]: I don't think we want a mux here. We want one per pull subscriber. Otherwise we will have issues.

File: js.go [hunk 4]
```diff
@@ -826,6 +863,67 @@ type jsSub struct {
...
+	sb.Write(b[:])
+	return sb.String()
+}
+
+// handleFetch is delivered a message requested by pull subscribers
+// when calling Fetch.
+func (jsi *jsSub) handleFetch(m *Msg) {
+	jsi.mu.Lock()
+	if len(jsi.freqs) > 0 {
+		mch := jsi.freqs[0]
```
Comment (derekcollison) [inline js.go:888]: This pattern will bite us ;)

**Review by derekcollison — CHANGES_REQUESTED** [review]

File: js.go (hunk 2 above)
Comment (wallyqs) [reply js.go:409]: yes it is one per pull subscribers, because they have their own queue of inflight requests

File: js.go (hunk 4 above)
Comment (wallyqs) [reply js.go:888]: I wonder if this would work better as a channel of channels...

Comment (derekcollison) [reply js.go:888]: Let's do the append trick we discussed.

Comment (derekcollison) [reply js.go:888]: And add to ping/pong as needed.

File: js.go [hunk 5]
```diff
@@ -400,6 +401,28 @@ func (paf *pubAckFuture) Msg() *Msg {
...
+func (js *js) pullSubscribe(subj string) (*Subscription, error) {
+	jsi := &jsSub{js: js, pull: true}
+
+	// Similar to async request handler we create a wildcard subscription for making requests,
+	// though we do not use the token based approach since we cannot match the response to
+	// the requestor due to JS subject being remapped on delivery.  Instead, we just use an array
+	// of channels similar to how ping/pong interval is handled and send the message to the first
+	// available requestor via a channel.
+	jsi.rr = rand.New(rand.NewSource(time.Now().UnixNano()))
+	jsi.rpre = fmt.Sprintf("%s.", NewInbox())
```
Comment (wallyqs) [inline js.go:415]: Changed to just use `nats.NewInbox` instead which should work for this case.

File: js.go (hunk 4 above)
Comment (wallyqs) [reply js.go]: Sending different PR for the ping/pong change

Comment (wallyqs) [reply js.go]: https://github.com/nats-io/nats.go/pull/723

File: js.go [hunk 6]
```diff
@@ -826,6 +857,67 @@ type jsSub struct {
...
+	sb.Write(b[:])
+	return sb.String()
+}
+
+// handleFetch is delivered a message requested by pull subscribers
+// when calling Fetch.
+func (jsi *jsSub) handleFetch(m *Msg) {
+	jsi.mu.Lock()
+	if len(jsi.freqs) == 0 {
+		jsi.mu.Unlock()
```
Comment (derekcollison) [inline js.go:882]: This something we should alert user about?

File: js.go [hunk 7]
```diff
@@ -826,6 +857,67 @@ type jsSub struct {
...
+		return
+	}
+	mch := jsi.freqs[0]
+	jsi.freqs = append(jsi.freqs[:0], jsi.freqs[1:]...)
```
Comment (derekcollison) [inline js.go:886]: Check if we have more than 1, if not optimize that case with a reslice to [:0]

**Review by derekcollison — CHANGES_REQUESTED** [review]

File: js.go (hunk 6 above)
Comment (wallyqs) [reply js.go:882]: In theory this should not happen (unless there is a race in the code), but maybe we should send an async error just in case it occurs

Comment (wallyqs) [reply js.go:891]: Added dispatching the error to the async error handler just in case it might happen

File: js.go (hunk 7 above)
Comment (wallyqs) [reply js.go]: Updated this one too

File: js.go [hunk 8]
```diff
@@ -400,6 +401,28 @@ func (paf *pubAckFuture) Msg() *Msg {
...
+	jsi := &jsSub{js: js, pull: true}
+
+	// Similar to async request handler we create a wildcard subscription for making requests,
+	// though we do not use the token based approach since we cannot match the response to
+	// the requestor due to JS subject being remapped on delivery.  Instead, we just use an array
+	// of channels similar to how ping/pong interval is handled and send the message to the first
+	// available requestor via a channel.
+	jsi.rr = rand.New(rand.NewSource(time.Now().UnixNano()))
+	jsi.rpre = fmt.Sprintf("%s.", NewInbox())
+	jsi.freqs = make([]chan *Msg, 0)
```
Comment (derekcollison) [inline js.go:416]: Would remove.

**Review by derekcollison — APPROVED** [review]
LGTM - one minor suggestion.

**Review by derekcollison — APPROVED** [review]
LGTM

**Pull request merged**